airbyte-integrations/connectors/source-google-ads/unit_tests/test_errors.py

```diff
@@ -0,0 +1,124 @@
+
+from .common import MockGoogleAdsClient, mock_google_ads_request_failure
+
+params = [
+    ("USER_PERMISSION_DENIED",
+     "Failed to access the customer '123'. Ensure the customer is linked to your manager account or check your permissions to access this customer account."),
+
+    ("CUSTOMER_NOT_FOUND",
+     "Failed to access the customer '123'. Ensure the customer is linked to your manager account or check your permissions to access this customer account."),
+
+    ("CUSTOMER_NOT_ENABLED",
+     ("The Google Ads account '123' you're trying to access is not enabled. "
+      "This may occur if the account hasn't been fully set up or activated. "
+      "Please ensure that you've completed all necessary setup steps in the Google Ads platform and that the account is active. "
+      "If the account is recently created, it might take some time before it's fully activated.")),
+
+    ("QUERY_ERROR", "Incorrect custom query. Error in query: unexpected end of query."),
+
+    ("RESOURCE_EXHAUSTED",
+     "You've exceeded your 24-hour quota limits for operations on your Google Ads account '123'. Try again later."),
+
+    ("UNEXPECTED_ERROR", "Unexpected error message")
+]
+
+
+@pytest.mark.parametrize(("exception", "error_message"), params)
+def test_expected_errors(mocker, config, exception, error_message):
+    mock_google_ads_request_failure(mocker, exception)
+    source = SourceGoogleAds()
+    with pytest.raises(AirbyteTracedException) as exception:
+        status_ok, error = source.check_connection(AirbyteLogger(), config)
+    assert exception.value.message == error_message
+
+
+@pytest.mark.parametrize(
+    ("cls", "raise_expected"),
+    (
+            (AdGroupLabels, False),
+            (Labels, False),
+            (ServiceAccounts, True),
+    ),
+)
+def test_read_record_error_handling(mocker, config, customers, cls, raise_expected):
+    mock_google_ads_request_failure(mocker, "CUSTOMER_NOT_ENABLED")
+    google_api = GoogleAds(credentials=config["credentials"])
+    stream = cls(api=google_api, customers=customers)
+    if raise_expected:
+        with pytest.raises(AirbyteTracedException) as exception:
+            for _ in stream.read_records(sync_mode=Mock(), stream_slice={"customer_id": "1234567890"}):
+                pass
+        assert exception.value.message == ("The Google Ads account '1234567890' you're trying to access is not enabled. "
+                                           "This may occur if the account hasn't been fully set up or activated. Please "
+                                           "ensure that you've completed all necessary setup steps in the Google Ads "
+                                           "platform and that the account is active. If the account is recently created, "
+                                           "it might take some time before it's fully activated.")
+    else:
+        for _ in stream.read_records(sync_mode=Mock(), stream_slice={"customer_id": "1234567890"}):
```

Comment:
nit: this is duplication from within the `with` context. You could use `from contextlib import nullcontext as does_not_raise` to avoid duplicating code

Reply:
Fixed